File: src/utils/aiEngine.ts
import { PieceValue, Position, PieceColor, Direction, ThreatType } from '../types';
import { 
  isValidPosition, 
  isFirstMove, 
  getMoveCount, 
  checkWin, 
  isCornerOrEdge, 
  getCenterNearMove 
} from './gameLogic';

// 高级AI引擎 - 基于Minimax + Alpha-Beta剪枝
export class AdvancedAIEngine {
  private static readonly WIN_SCORE = 100000;
  private static readonly LOSE_SCORE = -100000;
  private static readonly MAX_DEPTH = 4;
  private static readonly SEARCH_RADIUS = 2;
  
  // 开局库 - 前几步的固定走法
  private static readonly OPENING_BOOK = new Map<string, Position[]>([
    // 天元开局
    ['7,7', [{ row: 6, col: 6 }, { row: 8, col: 8 }, { row: 6, col: 8 }, { row: 8, col: 6 }]],
    // 边角开局
    ['0,0', [{ row: 7, col: 7 }, { row: 6, col: 6 }, { row: 8, col: 8 }]],
    ['0,14', [{ row: 7, col: 7 }, { row: 6, col: 8 }, { row: 8, col: 6 }]],
    ['14,0', [{ row: 7, col: 7 }, { row: 8, col: 6 }, { row: 6, col: 8 }]],
    ['14,14', [{ row: 7, col: 7 }, { row: 8, col: 8 }, { row: 6, col: 6 }]]
  ]);

  // 威胁级别评估函数
  static getThreatLevel(board: PieceValue[][], row: number, col: number, player: PieceValue): number {
    let maxThreat = 0;
    const directions: Direction[] = [
      { dx: 0, dy: 1 },   // 横向
      { dx: 1, dy: 0 },   // 纵向
      { dx: 1, dy: 1 },   // 主对角线
      { dx: 1, dy: -1 }   // 副对角线
    ];

    for (const dir of directions) {
      let count = 1;
      
      // 检查正方向
      for (let i = 1; i < 5; i++) {
        const newRow = row + dir.dx * i;
        const newCol = col + dir.dy * i;
        if (!isValidPosition(newRow, newCol) || board[newRow][newCol] !== player) {
          break;
        }
        count++;
      }

      // 检查反方向
      for (let i = 1; i < 5; i++) {
        const newRow = row - dir.dx * i;
        const newCol = col - dir.dy * i;
        if (!isValidPosition(newRow, newCol) || board[newRow][newCol] !== player) {
          break;
        }
        count++;
      }

      maxThreat = Math.max(maxThreat, count);
    }

    return maxThreat;
  }

  // 评估特定方向的威胁
  static evaluateDirection(board: PieceValue[][], row: number, col: number, dir: Direction, player: PieceValue): number {
    let score = 0;
    let consecutive = 1;
    let blocked = 0;

    // 检查正方向
    for (let i = 1; i < 5; i++) {
      const newRow = row + dir.dx * i;
      const newCol = col + dir.dy * i;
      
      if (!isValidPosition(newRow, newCol)) {
        blocked++;
        break;
      }
      
      if (board[newRow][newCol] === player) {
        consecutive++;
      } else if (board[newRow][newCol] === 0) {
        break;
      } else {
        blocked++;
        break;
      }
    }

    // 检查反方向
    for (let i = 1; i < 5; i++) {
      const newRow = row - dir.dx * i;
      const newCol = col - dir.dy * i;
      
      if (!isValidPosition(newRow, newCol)) {
        blocked++;
        break;
      }
      
      if (board[newRow][newCol] === player) {
        consecutive++;
      } else if (board[newRow][newCol] === 0) {
        break;
      } else {
        blocked++;
        break;
      }
    }

    // 根据连子数和阻挡情况评分
    if (consecutive >= 5) score += 10000;
    else if (consecutive === 4 && blocked === 0) score += 1000; // 活四
    else if (consecutive === 4 && blocked === 1) score += 100;  // 冲四
    else if (consecutive === 3 && blocked === 0) score += 100;  // 活三
    else if (consecutive === 3 && blocked === 1) score += 10;   // 眠三
    else if (consecutive === 2 && blocked === 0) score += 10;   // 活二
    else if (consecutive === 2 && blocked === 1) score += 1;    // 眠二

    return score;
  }

  // 检查是否有双四威胁（44）
  static hasDoubleFour(board: PieceValue[][], row: number, col: number, player: PieceValue): boolean {
    let fourCount = 0;
    const directions: Direction[] = [
      { dx: 0, dy: 1 }, { dx: 1, dy: 0 },
      { dx: 1, dy: 1 }, { dx: 1, dy: -1 }
    ];
    
    for (const dir of directions) {
      if (this.evaluateDirection(board, row, col, dir, player) >= 1000) { // 活四
        fourCount++;
      }
    }
    
    return fourCount >= 2;
  }

  // 检查是否有四三威胁（43）
  static hasFourThree(board: PieceValue[][], row: number, col: number, player: PieceValue): boolean {
    let fourCount = 0;
    let threeCount = 0;
    const directions: Direction[] = [
      { dx: 0, dy: 1 }, { dx: 1, dy: 0 },
      { dx: 1, dy: 1 }, { dx: 1, dy: -1 }
    ];
    
    for (const dir of directions) {
      const score = this.evaluateDirection(board, row, col, dir, player);
      if (score >= 1000) { // 活四
        fourCount++;
      } else if (score >= 100) { // 活三
        threeCount++;
      }
    }
    
    return fourCount >= 1 && threeCount >= 1;
  }

  // 检查是否有活三
  static hasLiveThree(board: PieceValue[][], row: number, col: number, player: PieceValue): boolean {
    const directions: Direction[] = [
      { dx: 0, dy: 1 }, { dx: 1, dy: 0 },
      { dx: 1, dy: 1 }, { dx: 1, dy: -1 }
    ];
    
    for (const dir of directions) {
      if (this.evaluateDirection(board, row, col, dir, player) >= 100) { // 活三
        return true;
      }
    }
    
    return false;
  }

  // 检查是否有双三威胁（33）
  static hasDoubleThree(board: PieceValue[][], row: number, col: number, player: PieceValue): boolean {
    let threeCount = 0;
    const directions: Direction[] = [
      { dx: 0, dy: 1 }, { dx: 1, dy: 0 },
      { dx: 1, dy: 1 }, { dx: 1, dy: -1 }
    ];
    
    for (const dir of directions) {
      if (this.evaluateDirection(board, row, col, dir, player) >= 100) { // 活三
        threeCount++;
      }
    }
    
    return threeCount >= 2;
  }

  // 评估位置价值
  static evaluatePosition(board: PieceValue[][], row: number, col: number, player: PieceValue): number {
    let score = 0;
    const directions: Direction[] = [
      { dx: 0, dy: 1 }, { dx: 1, dy: 0 },
      { dx: 1, dy: 1 }, { dx: 1, dy: -1 }
    ];
    
    for (const dir of directions) {
      score += this.evaluateDirection(board, row, col, dir, player);
    }
    
    return score;
  }

  // 获取所有可能的移动位置
  static getPossibleMoves(board: PieceValue[][], centerRow: number = 7, centerCol: number = 7): Position[] {
    const moves: Position[] = [];
    const radius = this.SEARCH_RADIUS;
    
    for (let row = Math.max(0, centerRow - radius); row <= Math.min(14, centerRow + radius); row++) {
      for (let col = Math.max(0, centerCol - radius); col <= Math.min(14, centerCol + radius); col++) {
        if (board[row][col] === 0) {
          // 检查周围是否有棋子
          let hasNeighbor = false;
          for (let dr = -1; dr <= 1; dr++) {
            for (let dc = -1; dc <= 1; dc++) {
              const newRow = row + dr;
              const newCol = col + dc;
              if (isValidPosition(newRow, newCol) && board[newRow][newCol] !== 0) {
                hasNeighbor = true;
                break;
              }
            }
            if (hasNeighbor) break;
          }
          if (hasNeighbor) {
            moves.push({ row, col });
          }
        }
      }
    }
    
    return moves;
  }

  // 评估棋盘状态
  static evaluateBoard(board: PieceValue[][], player: PieceValue): number {
    let score = 0;
    
    // 遍历所有位置，评估每个位置的威胁
    for (let row = 0; row < 15; row++) {
      for (let col = 0; col < 15; col++) {
        if (board[row][col] === player) {
          score += this.evaluatePosition(board, row, col, player);
        } else if (board[row][col] === (-player as PieceValue)) {
          score -= this.evaluatePosition(board, row, col, (-player as PieceValue));
        }
      }
    }
    
    return score;
  }

  // Minimax算法配合Alpha-Beta剪枝
  static minimax(
    board: PieceValue[][], 
    depth: number, 
    alpha: number, 
    beta: number, 
    isMaximizing: boolean, 
    player: PieceValue,
    opponent: PieceValue
  ): number {
    // 检查游戏结束条件
    if (depth === 0) {
      return this.evaluateBoard(board, player);
    }

    // 检查是否有获胜
    for (let row = 0; row < 15; row++) {
      for (let col = 0; col < 15; col++) {
        if (board[row][col] === 0) {
          board[row][col] = isMaximizing ? player : opponent;
          if (checkWin(board, row, col)) {
            board[row][col] = 0;
            return isMaximizing ? this.WIN_SCORE : this.LOSE_SCORE;
          }
          board[row][col] = 0;
        }
      }
    }

    const moves = this.getPossibleMoves(board);
    if (moves.length === 0) {
      return this.evaluateBoard(board, player);
    }

    if (isMaximizing) {
      let maxEval = -Infinity;
      for (const move of moves) {
        board[move.row][move.col] = player;
        const evaluation = this.minimax(board, depth - 1, alpha, beta, false, player, opponent);
        board[move.row][move.col] = 0;
        maxEval = Math.max(maxEval, evaluation);
        alpha = Math.max(alpha, evaluation);
        if (beta <= alpha) break; // Alpha-Beta剪枝
      }
      return maxEval;
    } else {
      let minEval = Infinity;
      for (const move of moves) {
        board[move.row][move.col] = opponent;
        const evaluation = this.minimax(board, depth - 1, alpha, beta, true, player, opponent);
        board[move.row][move.col] = 0;
        minEval = Math.min(minEval, evaluation);
        beta = Math.min(beta, evaluation);
        if (beta <= alpha) break; // Alpha-Beta剪枝
      }
      return minEval;
    }
  }

  // 检查威胁级别 - 新增威胁检测函数
  static checkThreatLevel(board: PieceValue[][], row: number, col: number, player: PieceValue): number {
    const directions: Direction[] = [
      { dx: 0, dy: 1 },   // 横向
      { dx: 1, dy: 0 },   // 纵向
      { dx: 1, dy: 1 },   // 主对角线
      { dx: 1, dy: -1 }   // 副对角线
    ];

    let maxThreat = 0;
    for (const dir of directions) {
      let count = 1;
      let blocked = 0;
      
      // 检查正方向
      for (let i = 1; i < 5; i++) {
        const newRow = row + dir.dx * i;
        const newCol = col + dir.dy * i;
        if (!isValidPosition(newRow, newCol)) {
          blocked++;
          break;
        }
        if (board[newRow][newCol] === player) {
          count++;
        } else if (board[newRow][newCol] === 0) {
          break;
        } else {
          blocked++;
          break;
        }
      }

      // 检查反方向
      for (let i = 1; i < 5; i++) {
        const newRow = row - dir.dx * i;
        const newCol = col - dir.dy * i;
        if (!isValidPosition(newRow, newCol)) {
          blocked++;
          break;
        }
        if (board[newRow][newCol] === player) {
          count++;
        } else if (board[newRow][newCol] === 0) {
          break;
        } else {
          blocked++;
          break;
        }
      }

      // 威胁评分：连子数越多，威胁越大
      if (count >= 5) maxThreat = Math.max(maxThreat, 10000); // 五连
      else if (count === 4 && blocked === 0) maxThreat = Math.max(maxThreat, 5000); // 活四
      else if (count === 4 && blocked === 1) maxThreat = Math.max(maxThreat, 1000); // 冲四
      else if (count === 3 && blocked === 0) maxThreat = Math.max(maxThreat, 500);  // 活三
      else if (count === 3 && blocked === 1) maxThreat = Math.max(maxThreat, 100);  // 眠三
      else if (count === 2 && blocked === 0) maxThreat = Math.max(maxThreat, 50);   // 活二
    }

    return maxThreat;
  }

  // 获取所有威胁位置
  static getThreatPositions(board: PieceValue[][], player: PieceValue): Position[] {
    const threats: Position[] = [];
    
    for (let row = 0; row < 15; row++) {
      for (let col = 0; col < 15; col++) {
        if (board[row][col] === 0) {
          const threatLevel = this.checkThreatLevel(board, row, col, player);
          if (threatLevel >= 100) { // 只考虑活三以上的威胁
            threats.push({ row, col, threatLevel } as Position & { threatLevel: number });
          }
        }
      }
    }
    
    // 按威胁级别排序
    return threats.sort((a, b) => (b as any).threatLevel - (a as any).threatLevel);
  }

  // 获取最佳移动
  static getBestMove(board: PieceValue[][], aiPlayer: PieceValue, humanPlayer: PieceValue): Position | null {
    // 检查开局库
    const moveCount = getMoveCount(board);
    if (moveCount <= 4) {
      const openingMove = this.getOpeningMove(board, aiPlayer, humanPlayer);
      if (openingMove) {
        return openingMove;
      }
    }

    // 1. 优先检查AI立即获胜
    for (let row = 0; row < 15; row++) {
      for (let col = 0; col < 15; col++) {
        if (board[row][col] === 0) {
          board[row][col] = aiPlayer;
          if (checkWin(board, row, col)) {
            board[row][col] = 0;
            return { row, col };
          }
          board[row][col] = 0;
        }
      }
    }

    // 2. 优先检查阻止对手立即获胜
    for (let row = 0; row < 15; row++) {
      for (let col = 0; col < 15; col++) {
        if (board[row][col] === 0) {
          board[row][col] = humanPlayer;
          if (checkWin(board, row, col)) {
            board[row][col] = 0;
            return { row, col };
          }
          board[row][col] = 0;
        }
      }
    }

    // 3. 优势权衡：评估攻击和防守的优先级
    const aiThreats = this.getThreatPositions(board, aiPlayer);
    const humanThreats = this.getThreatPositions(board, humanPlayer);
    
    // 计算双方威胁强度
    const aiThreatStrength = this.calculateThreatStrength(aiThreats);
    const humanThreatStrength = this.calculateThreatStrength(humanThreats);
    
    // 计算当前局面优势
    const currentAdvantage = this.calculateCurrentAdvantage(board, aiPlayer, humanPlayer);
    
    // 决策逻辑：权衡攻击和防守
    const shouldAttack = this.shouldPrioritizeAttack(aiThreatStrength, humanThreatStrength, currentAdvantage);
    
    // 只在开发环境下输出调试日志
    if (process.env.NODE_ENV === 'development') {
      console.log('🤖 AI决策分析:', {
        aiThreats: aiThreats.length,
        humanThreats: humanThreats.length,
        aiThreatStrength,
        humanThreatStrength,
        currentAdvantage,
        shouldAttack
      });
    }
    
    if (shouldAttack && aiThreats.length > 0) {
      // 优先进攻：我方威胁大于敌方威胁，或者我方有明显优势
      if (process.env.NODE_ENV === 'development') {
        console.log('⚔️ 选择进攻策略');
      }
      return aiThreats[0];
    } else if (humanThreats.length > 0) {
      // 优先防守：敌方威胁较大，或者我方优势不明显
      if (process.env.NODE_ENV === 'development') {
        console.log('🛡️ 选择防守策略');
      }
      return humanThreats[0];
    }

    // 4. 使用Minimax算法进行深度搜索
    const moves = this.getPossibleMoves(board);
    let bestMove: Position | null = null;
    let bestScore = -Infinity;

    for (const move of moves) {
      board[move.row][move.col] = aiPlayer;
      const score = this.minimax(board, this.MAX_DEPTH, -Infinity, Infinity, false, aiPlayer, humanPlayer);
      board[move.row][move.col] = 0;
      
      if (score > bestScore) {
        bestScore = score;
        bestMove = move;
      }
    }

    return bestMove;
  }

  // 计算威胁强度
  private static calculateThreatStrength(threats: Position[]): number {
    if (threats.length === 0) return 0;
    
    let totalStrength = 0;
    for (const threat of threats) {
      // 根据威胁级别计算强度
      const threatLevel = (threat as any).threatLevel || 0;
      totalStrength += threatLevel;
    }
    
    return totalStrength;
  }

  // 计算当前局面优势
  private static calculateCurrentAdvantage(board: PieceValue[][], aiPlayer: PieceValue, humanPlayer: PieceValue): number {
    let aiScore = 0;
    let humanScore = 0;
    
    // 评估每个位置的威胁
    for (let row = 0; row < 15; row++) {
      for (let col = 0; col < 15; col++) {
        if (board[row][col] === aiPlayer) {
          aiScore += this.evaluatePosition(board, row, col, aiPlayer);
        } else if (board[row][col] === humanPlayer) {
          humanScore += this.evaluatePosition(board, row, col, humanPlayer);
        }
      }
    }
    
    return aiScore - humanScore;
  }

  // 判断是否应该优先进攻
  private static shouldPrioritizeAttack(aiThreatStrength: number, humanThreatStrength: number, currentAdvantage: number): boolean {
    // 如果我方威胁明显大于敌方威胁，优先进攻
    if (aiThreatStrength > humanThreatStrength * 1.5) {
      return true;
    }
    
    // 如果我方有明显优势（分数差大于1000），优先进攻
    if (currentAdvantage > 1000) {
      return true;
    }
    
    // 如果敌方威胁很小（小于500），优先进攻
    if (humanThreatStrength < 500) {
      return true;
    }
    
    // 如果双方威胁相当，但我方优势大于500，优先进攻
    if (Math.abs(aiThreatStrength - humanThreatStrength) < 200 && currentAdvantage > 500) {
      return true;
    }
    
    // 其他情况优先防守
    return false;
  }

  // 获取开局移动
  private static getOpeningMove(board: PieceValue[][], aiPlayer: PieceValue, humanPlayer: PieceValue): Position | null {
    const moveCount = getMoveCount(board);
    
    if (moveCount === 0) {
      return { row: 7, col: 7 }; // 天元
    }
    
    if (moveCount === 1) {
      // 找到对手的第一步
      for (let row = 0; row < 15; row++) {
        for (let col = 0; col < 15; col++) {
          if (board[row][col] === humanPlayer) {
            const key = `${row},${col}`;
            const responses = this.OPENING_BOOK.get(key);
            if (responses && responses.length > 0) {
              return responses[0];
            }
          }
        }
      }
    }
    
    return null;
  }
}

// 为了保持向后兼容，导出原有的函数
export const getThreatLevel = AdvancedAIEngine.getThreatLevel;
export const evaluateDirection = AdvancedAIEngine.evaluateDirection;
export const hasDoubleFour = AdvancedAIEngine.hasDoubleFour;
export const hasFourThree = AdvancedAIEngine.hasFourThree;
export const hasLiveThree = AdvancedAIEngine.hasLiveThree;
export const hasDoubleThree = AdvancedAIEngine.hasDoubleThree;
export const evaluatePosition = AdvancedAIEngine.evaluatePosition;

File: src-tauri/src/ai_engine.rs
// AI引擎 - Rust实现
// 高性能的五子棋AI算法

const BOARD_SIZE: usize = 15;
const WIN_LENGTH: usize = 5;
const MAX_DEPTH: i32 = 4;
const WIN_SCORE: i32 = 100000;
const LOSE_SCORE: i32 = -100000;

// 方向向量
const DIRECTIONS: [(i32, i32); 4] = [
    (0, 1),   // 横向
    (1, 0),   // 纵向
    (1, 1),   // 主对角线
    (1, -1),  // 副对角线
];

// 检查位置是否有效
fn is_valid_position(row: i32, col: i32) -> bool {
    row >= 0 && row < BOARD_SIZE as i32 && col >= 0 && col < BOARD_SIZE as i32
}

// 检查是否获胜
pub fn check_win(board: &Vec<Vec<i32>>, row: usize, col: usize) -> bool {
    let player = board[row][col];
    if player == 0 {
        return false;
    }

    for (dx, dy) in DIRECTIONS.iter() {
        let mut count = 1;
        
        // 检查正方向
        for i in 1..WIN_LENGTH {
            let new_row = row as i32 + dx * i as i32;
            let new_col = col as i32 + dy * i as i32;
            if !is_valid_position(new_row, new_col) || 
               board[new_row as usize][new_col as usize] != player {
                break;
            }
            count += 1;
        }
        
        // 检查反方向
        for i in 1..WIN_LENGTH {
            let new_row = row as i32 - dx * i as i32;
            let new_col = col as i32 - dy * i as i32;
            if !is_valid_position(new_row, new_col) || 
               board[new_row as usize][new_col as usize] != player {
                break;
            }
            count += 1;
        }
        
        if count >= WIN_LENGTH {
            return true;
        }
    }
    
    false
}

// 评估位置分数
fn evaluate_position(board: &Vec<Vec<i32>>, row: usize, col: usize, player: i32) -> i32 {
    let mut score = 0;
    
    for (dx, dy) in DIRECTIONS.iter() {
        let mut count = 1;
        let mut blocked = 0;
        
        // 检查正方向
        for i in 1..WIN_LENGTH {
            let new_row = row as i32 + dx * i as i32;
            let new_col = col as i32 + dy * i as i32;
            if !is_valid_position(new_row, new_col) {
                blocked += 1;
                break;
            }
            let cell = board[new_row as usize][new_col as usize];
            if cell == player {
                count += 1;
            } else if cell == 0 {
                break;
            } else {
                blocked += 1;
                break;
            }
        }
        
        // 检查反方向
        for i in 1..WIN_LENGTH {
            let new_row = row as i32 - dx * i as i32;
            let new_col = col as i32 - dy * i as i32;
            if !is_valid_position(new_row, new_col) {
                blocked += 1;
                break;
            }
            let cell = board[new_row as usize][new_col as usize];
            if cell == player {
                count += 1;
            } else if cell == 0 {
                break;
            } else {
                blocked += 1;
                break;
            }
        }
        
        // 根据连子数量和阻塞情况评分
        if count >= WIN_LENGTH {
            score += WIN_SCORE;
        } else if count == WIN_LENGTH - 1 && blocked == 0 {
            score += 5000; // 活四
        } else if count == WIN_LENGTH - 1 && blocked == 1 {
            score += 1000; // 冲四
        } else if count == WIN_LENGTH - 2 && blocked == 0 {
            score += 500;  // 活三
        } else if count == WIN_LENGTH - 2 && blocked == 1 {
            score += 100;  // 眠三
        } else if count == WIN_LENGTH - 3 && blocked == 0 {
            score += 50;   // 活二
        } else if count == WIN_LENGTH - 3 && blocked == 1 {
            score += 10;   // 眠二
        }
    }
    
    score
}

// 评估整个棋盘
pub fn evaluate_board(board: &Vec<Vec<i32>>, player: i32) -> i32 {
    let mut score = 0;
    
    for row in 0..BOARD_SIZE {
        for col in 0..BOARD_SIZE {
            if board[row][col] == player {
                score += evaluate_position(board, row, col, player);
            } else if board[row][col] == -player {
                score -= evaluate_position(board, row, col, -player);
            }
        }
    }
    
    score
}

// 获取可能的移动位置
fn get_possible_moves(board: &Vec<Vec<i32>>) -> Vec<(usize, usize)> {
    let mut moves = Vec::new();
    
    // 检查是否为空棋盘
    let mut has_pieces = false;
    for row in 0..BOARD_SIZE {
        for col in 0..BOARD_SIZE {
            if board[row][col] != 0 {
                has_pieces = true;
                break;
            }
        }
        if has_pieces { break; }
    }
    
    // 如果棋盘为空，返回中心位置
    if !has_pieces {
        moves.push((7, 7));
        return moves;
    }
    
    // 找到所有已有棋子的位置
    let mut piece_positions = Vec::new();
    for row in 0..BOARD_SIZE {
        for col in 0..BOARD_SIZE {
            if board[row][col] != 0 {
                piece_positions.push((row, col));
            }
        }
    }
    
    // 为每个已有棋子周围的空位添加候选位置
    let mut candidate_positions = std::collections::HashSet::new();
    
    for (piece_row, piece_col) in piece_positions {
        for dr in -2..=2 {
            for dc in -2..=2 {
                let new_row = piece_row as i32 + dr;
                let new_col = piece_col as i32 + dc;
                
                if is_valid_position(new_row, new_col) {
                    let row = new_row as usize;
                    let col = new_col as usize;
                    
                    if board[row][col] == 0 {
                        candidate_positions.insert((row, col));
                    }
                }
            }
        }
    }
    
    // 将候选位置添加到移动列表
    for (row, col) in candidate_positions {
        moves.push((row, col));
    }
    
    // 如果没有找到任何移动，返回所有空位
    if moves.is_empty() {
        for row in 0..BOARD_SIZE {
            for col in 0..BOARD_SIZE {
                if board[row][col] == 0 {
                    moves.push((row, col));
                }
            }
        }
    }
    
    moves
}

// Minimax算法
fn minimax(
    board: &mut Vec<Vec<i32>>,
    depth: i32,
    alpha: i32,
    beta: i32,
    is_maximizing: bool,
    player: i32,
    opponent: i32,
) -> i32 {
    if depth == 0 {
        return evaluate_board(board, player);
    }
    
    // 检查获胜
    for row in 0..BOARD_SIZE {
        for col in 0..BOARD_SIZE {
            if board[row][col] == 0 {
                board[row][col] = if is_maximizing { player } else { opponent };
                if check_win(board, row, col) {
                    board[row][col] = 0;
                    return if is_maximizing { WIN_SCORE } else { LOSE_SCORE };
                }
                board[row][col] = 0;
            }
        }
    }
    
    let moves = get_possible_moves(board);
    if moves.is_empty() {
        return evaluate_board(board, player);
    }
    
    if is_maximizing {
        let mut max_eval = LOSE_SCORE;
        let mut alpha = alpha;
        for (row, col) in moves {
            board[row][col] = player;
            let eval = minimax(board, depth - 1, alpha, beta, false, player, opponent);
            board[row][col] = 0;
            max_eval = max_eval.max(eval);
            alpha = alpha.max(eval);
            if beta <= alpha {
                break;
            }
        }
        max_eval
    } else {
        let mut min_eval = WIN_SCORE;
        let mut beta = beta;
        for (row, col) in moves {
            board[row][col] = opponent;
            let eval = minimax(board, depth - 1, alpha, beta, true, player, opponent);
            board[row][col] = 0;
            min_eval = min_eval.min(eval);
            beta = beta.min(eval);
            if beta <= alpha {
                break;
            }
        }
        min_eval
    }
}

// 获取最佳移动
pub fn get_best_move(board: &Vec<Vec<i32>>, ai_player: i32, human_player: i32) -> Option<(usize, usize)> {
    println!("🤖 Rust AI开始思考... ai_player: {}, human_player: {}", ai_player, human_player);
    
    // 检查AI立即获胜
    for row in 0..BOARD_SIZE {
        for col in 0..BOARD_SIZE {
            if board[row][col] == 0 {
                let mut test_board = board.clone();
                test_board[row][col] = ai_player;
                if check_win(&test_board, row, col) {
                    println!("🎯 AI立即获胜: ({}, {})", row, col);
                    return Some((row, col));
                }
            }
        }
    }
    
    // 检查阻止对手立即获胜
    for row in 0..BOARD_SIZE {
        for col in 0..BOARD_SIZE {
            if board[row][col] == 0 {
                let mut test_board = board.clone();
                test_board[row][col] = human_player;
                if check_win(&test_board, row, col) {
                    println!("🛡️ 阻止对手获胜: ({}, {})", row, col);
                    return Some((row, col));
                }
            }
        }
    }
    
    // 使用Minimax算法
    let moves = get_possible_moves(board);
    println!("📋 候选移动数量: {}", moves.len());
    
    if moves.is_empty() {
        println!("⚠️ 没有可用移动");
        return None;
    }
    
    let mut best_move = None;
    let mut best_score = LOSE_SCORE;
    
    for (row, col) in moves {
        let mut test_board = board.clone();
        test_board[row][col] = ai_player;
        let score = minimax(&mut test_board, MAX_DEPTH, LOSE_SCORE, WIN_SCORE, false, ai_player, human_player);
        
        println!("📍 位置 ({}, {}) 得分: {}", row, col, score);
        
        if score > best_score {
            best_score = score;
            best_move = Some((row, col));
        }
    }
    
    if let Some((row, col)) = best_move {
        println!("✅ 最佳移动: ({}, {}) 得分: {}", row, col, best_score);
    }
    
    best_move
}
